### ISSUE TITLE
Add Cnn or DenseLayer, LSTM cannot use TruncatedBPTT

### ISSUE THREAD

**Issue by longzhendong** [deeplearning4j/deeplearning4j#6382]
Add Cnn or DenseLayer, LSTM cannot use TruncatedBPTT，Add Cnn or DenseLayer, LSTM cannot use TruncatedBPTT, my network structure is as follows：

final GraphBuilder graphBuilder = builder.graphBuilder().pretrain(false).backprop(true)
				.backpropType(BackpropType.TruncatedBPTT)
				.tBPTTBackwardLength(5)
				.tBPTTForwardLength(5)
				.addInputs("inputLine")
				.setInputTypes(InputType.recurrent(501 * 501))
				.addLayer("cnn1", new ConvolutionLayer.Builder(6, 6)
                        .nIn(1) 
                        .nOut(15)
                        .stride(3, 3)
                        .activation(Activation.LEAKYRELU)
                        .weightInit(WeightInit.RELU)
                        .build(), "inputLine")//Output: (501-6+0)/3+1 = 166 -> 166*166*15
				.addLayer("pooling1", new SubsamplingLayer.Builder(SubsamplingLayer.PoolingType.MAX)
                        .kernelSize(4, 4)
                        .stride(2, 2).build(),"cnn1")//(166-4+0)/2+1 = 82
				.addLayer("cnn2", new ConvolutionLayer.Builder(4, 4)
                        .nIn(15)
                        .nOut(40)
                        .stride(2, 2)
                        .activation(Activation.LEAKYRELU)
                       .weightInit(WeightInit.RELU)
                        .build(), "pooling1")//Output: (82-4+0)/2+1 = 40 -> 40*40*40
				.addLayer("pooling2", new SubsamplingLayer.Builder(SubsamplingLayer.PoolingType.MAX)
                        .kernelSize(4, 4)
                        .stride(2, 2).build(),"cnn2")//(40-4+0)/2+1 = 19->19*19*40
				.addLayer("cnn3", new ConvolutionLayer.Builder(3, 3)
                        .nIn(40)
                        .nOut(64)
                        .stride(2, 2)
                        .activation(Activation.LEAKYRELU)
                        .weightInit(WeightInit.RELU)
                        .build(), "pooling2")//Output: (19-3+0)/2+1 = 9 -> 9*9*64
				.addLayer("fullConn",
						new DenseLayer.Builder()
		                .activation(Activation.RELU)
		                .nIn(9*9*64)
		                .nOut(HIDDEN_LAYER_WIDTH)
		                //.weightInit(WeightInit.RELU)
		                //.gradientNormalization(GradientNormalization.ClipElementWiseAbsoluteValue)
		                //.gradientNormalizationThreshold(10)
		                .build(),
						"cnn3")
				.addLayer("lstm",
						new LSTM.Builder().nIn( 9*9*64).nOut(128).activation(Activation.TANH)
						//.gradientNormalization(GradientNormalization.ClipElementWiseAbsoluteValue)
						//.gradientNormalizationThreshold(10)
						.dropOut(0.65)
						.build(),
						"fullConn")
				.addLayer("output",
						new RnnOutputLayer.Builder().nIn(128).nOut(501*501)  
								.activation(Activation.IDENTITY).lossFunction(LossFunctions.LossFunction.MSE)						
								//.gradientNormalization(GradientNormalization.ClipElementWiseAbsoluteValue)
								//.gradientNormalizationThreshold(10)//.weightInit(WeightInit.ZERO)
								//.updater(new RmsProp.Builder().learningRate(0.0005).build())
								//.updater(new Adam.Builder().learningRate(0.0005).build())
								.build(),
						"lstm")
				.setOutputs("output").inputPreProcessor("cnn1", new RnnToCnnPreProcessor(501, 501, 1))
				.inputPreProcessor("lstm", new CnnToRnnPreProcessor(9, 9, 64)) 
				.inputPreProcessor("fullConn", new CnnToFeedForwardPreProcessor(9, 9, 64))
				.inputPreProcessor("lstm", new FeedForwardToRnnPreProcessor());
		ComputationGraph net = new ComputationGraph(graphBuilder.build());

The stack of exceptions thrown is as follows：

Exception in thread "main" java.lang.IllegalStateException: 3D input expected to RNN layer expected, got 2
	at org.nd4j.base.Preconditions.throwStateEx(Preconditions.java:626)
	at org.nd4j.base.Preconditions.checkState(Preconditions.java:253)
	at org.deeplearning4j.nn.layers.recurrent.LSTM.activateHelper(LSTM.java:146)
	at org.deeplearning4j.nn.layers.recurrent.LSTM.rnnActivateUsingStoredState(LSTM.java:217)
	at org.deeplearning4j.nn.graph.ComputationGraph.ffToLayerActivationsInWS(ComputationGraph.java:2056)
	at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:1357)
	at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:1326)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:160)
	at org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(StochasticGradientDescent.java:51)
	at org.deeplearning4j.optimize.Solver.optimize(Solver.java:52)
	at org.deeplearning4j.nn.graph.ComputationGraph.doTruncatedBPTT(ComputationGraph.java:3534)
	at org.deeplearning4j.nn.graph.ComputationGraph.fitHelper(ComputationGraph.java:1140)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1098)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1065)

I guess the parameter validation is after PreProcessor

**Comment by jpseixasesilva** [comment]
Hello @longzhendong, can you solve the problem (Exception in thread "main" java.lang.IllegalStateException: 3D input expected to RNN layer expected, got 2) ?

**Comment by longzhendong** [comment]
@jpseixasesilva I haven't solved this problem yet
